astropy/wcs/fitswcs_low_level_api.py

```diff
@@ -0,0 +1,195 @@
+
+    @property
+    def world_axis_physical_types(self):
+        types = []
+        for axis_type in self._wcs.axis_type_names:
+            if axis_type.startswith('UT('):
+                types.append('time')
+            else:
+                types.append(CTYPE_TO_UCD1.get(axis_type, None))
+        return types
+
+    @property
+    def world_axis_units(self):
+        units = []
+        for unit in self._wcs.wcs.cunit:
+            if unit is None:
+                unit = ''
+            elif isinstance(unit, u.Unit):
+                unit = unit.to_string(format='vounit')
+            else:
+                try:
+                    unit = u.Unit(unit).to_string(format='vounit')
+                except u.UnitsError:
+                    unit = ''
+            units.append(unit)
+        return units
+
+    @property
+    def axis_correlation_matrix(self):
+
+        # If there are any distortions present, we assume that there may be
+        # correlations between all axes. Maybe if some distortions only apply
+        # to the image plane we can improve this?
+        for distortion_attribute in ('sip', 'det2im1', 'det2im2'):
+            if getattr(self._wcs, distortion_attribute):
+                return np.ones((self.n_world, self.n_pixel), dtype=bool)
+
+        # Assuming linear world coordinates along each axis, the correlation
+        # matrix would be given by whether or not the PC matrix is zero
+        matrix = self._wcs.wcs.get_pc() != 0
+
+        # We now need to check specifically for celestial coordinates since
+        # these can assume correlations because of spherical distortions. For
+        # each celestial coordinate we copy over the pixel dependencies from
+        # the other celestial coordinates.
+        celestial = (self._wcs.wcs.axis_types // 1000) % 10 == 2
+        celestial_indices = np.nonzero(celestial)[0]
+        for world1 in celestial_indices:
+            for world2 in celestial_indices:
+                if world1 != world2:
+                    matrix[world1] |= matrix[world2]
+                    matrix[world2] |= matrix[world1]
+
+        return matrix
+
+    def pixel_to_world_values(self, *pixel_arrays):
+        return self._wcs.all_pix2world(*pixel_arrays, 0)
+
+    def numpy_index_to_world_values(self, *indices):
+        return self._wcs.all_pix2world(*indices[::-1], 0)
+
+    def world_to_pixel_values(self, *world_arrays):
+        return self._wcs.all_world2pix(*world_arrays, 0)
+
+    def world_to_numpy_index_values(self, *world_arrays):
+        return np.round(self._wcs.all_world2pix(*world_arrays, 0)[::-1]).astype(int)
+
+    @property
+    def world_axis_object_components(self):
+        return _get_components_and_classes(self._wcs)[0]
+
+    @property
+    def world_axis_object_classes(self):
+        return _get_components_and_classes(self._wcs)[1]
+
+    @property
+    def serialized_classes(self):
+        return False
+
+
+def _get_components_and_classes(wcs):
```

Comment:
Also this could do with a docstring, at the very least to disambiguate the input value between `astropy.wcs.WCS` and the low level class.

Reply:
Done